perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_user_attribute_def_def_login_namespace_surf_ram_persistent_shadow.java

```diff
@@ -0,0 +1,92 @@
+
+	private final static Logger log = LoggerFactory.getLogger(urn_perun_user_attribute_def_def_login_namespace_surf_ram_persistent_shadow.class);
+	private final static String extSourceNameSurfRam = "https://proxy.acc.surf-ram.eduteams.org/proxy";
+	private final static String domainNameSurfRam = "@acc.surf-ram.eduteams.org";
+	private final static String attrNameSurfRam = "login-namespace:surf-ram-persistent-shadow";
+
+	/**
+	 * Filling implemented for login:namespace:surf-ram-persistent attribute
+	 * Format is: "[hash]@domainNameSurfRam" where [hash] represents sha1hash counted from user's id and perun instance id a login-namespace name
+	 *
+	 * @param perunSession PerunSession
+	 * @param user User to fill attribute for
+	 * @param attribute Attribute to fill value to
+	 * @return Filled attribute
+	 * @throws InternalErrorException
+	 */
+	@Override
+	public Attribute fillAttribute(PerunSessionImpl perunSession, User user, AttributeDefinition attribute) throws InternalErrorException {
+
+		Attribute filledAttribute = new Attribute(attribute);
+
+		if (attribute.getFriendlyName().equals(attrNameSurfRam)) {
+			filledAttribute.setValue(sha1HashCount(user, domainNameSurfRam).toString() + domainNameSurfRam);
+			return filledAttribute;
+		} else {
+			// without value
+			return filledAttribute;
+		}
+	}
+
+	/**
+	 * ChangedAttributeHook() sets UserExtSource with following properties:
+	 *  - extSourceType is IdP
+	 *  - extSourceName is stored in extSourceNameSurfRam
+	 *  - user's extSource login is the same as his surf-ram-persistent attribute
+	 *
+	 * @param session PerunSession
+	 * @param user User to set UserExtSource for
+	 * @param attribute Attribute containing SURF RAM ID
+	 * @throws cz.metacentrum.perun.core.api.exceptions.InternalErrorException
+	 */
+	@Override
+	public void changedAttributeHook(PerunSessionImpl session, User user, Attribute attribute) throws InternalErrorException {
+		try {
+			String userNamespace = attribute.getFriendlyNameParameter();
+
+			if(userNamespace.equals("surf-ram-persistent-shadow") && attribute.getValue() != null){
+				ExtSource extSource = session.getPerunBl().getExtSourcesManagerBl().getExtSourceByName(session, extSourceNameSurfRam);
+				UserExtSource userExtSource = new UserExtSource(extSource, 0, attribute.getValue().toString());
+
+				session.getPerunBl().getUsersManagerBl().addUserExtSource(session, user, userExtSource);
+			}
+		} catch (UserExtSourceExistsException ex) {
+			log.warn("SurfRam IdP external source already exists for the user.", ex);
+		} catch (ExtSourceNotExistsException ex) {
+			throw new InternalErrorException("IdP external source for SURF RAM doesn't exist.", ex);
+		}
+	}
+
+	public AttributeDefinition getAttributeDefinition() {
+		AttributeDefinition attr = new AttributeDefinition();
+		attr.setNamespace(AttributesManager.NS_USER_ATTR_DEF);
+		attr.setFriendlyName("login-namespace:surf-ram-persistent-shadow");
+		attr.setDisplayName("FENIX login");
```

Comment:
Probably should be ``SURF RAM login``, same for description.

Reply:
Cup&paste error. Thanks for spotting it.